Handle additional ISO 8601 format strings

From the 2020-08-18 daily test run:
- [TestINEGI.test_endpoints[dataflow] failed](https://github.com/khaeru/sdmx/runs/996789450?check_suite_focus=true#step:8:371):
  ```
  >       reader.push(elem, datetime.fromisoformat(text))
  E       ValueError: Invalid isoformat string: '2020-08-18T00:14:31.59849+05:00'

  sdmx/reader/sdmxml.py:720: ValueError
  ```
- [TestUNSD.test_endpoints[codelist] failed](https://github.com/khaeru/sdmx/runs/996789450?check_suite_focus=true#step:8:527):
  ```
  >       reader.push(elem, datetime.fromisoformat(text))
  E       ValueError: Invalid isoformat string: '2020-08-18T07:21:00.70664+02:00'

  sdmx/reader/sdmxml.py:720: ValueError
  ```

These dates are in valid [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601#Times), but Python's `datetime.fromisoformat()` ([docs](https://docs.python.org/3/library/datetime.html#datetime.datetime.fromisoformat)) only supports 3 or 6 decimal places for the seconds component of the time. These have 5, hence the errors.

Solutions:
- Extend the set of regexes.
- As suggested by the Python standard docs, add `dateutil` as a dependency to handle these.

--- a/sdmx/reader/sdmxml.py
+++ b/sdmx/reader/sdmxml.py
@@ -10,11 +10,11 @@ import logging
 import re
 from collections import defaultdict
 from copy import copy
-from datetime import datetime
 from itertools import chain, product
 from operator import itemgetter
 from sys import maxsize
 
+from dateutil.parser import isoparse
 from lxml import etree
 from lxml.etree import QName
 
@@ -702,22 +702,11 @@ def _text(reader, elem):
 
 @end("mes:Extracted mes:Prepared mes:ReportingBegin mes:ReportingEnd")
 def _datetime(reader, elem):
-    localname = QName(elem).localname
-
-    # Handle non-ISO date-times occuring in common messages
-    text, n = re.subn(r"(.*)Z$", r"\1+00:00", elem.text)
+    text, n = re.subn(r"(.*\.)(\d{6})\d+(\+.*)", r"\1\2\3", elem.text)
     if n > 0:
-        log.debug(f"Replace non-ISO 'Z' with '+00:00' in <{localname}>")
+        log.debug(f"Truncate sub-microsecond time in <{QName(elem).localname}>")
 
-    text, n = re.subn(r"(.*)-(\d{2}:\d{2})$", r"\1+\2", text)
-    if n > 0:
-        log.debug(f"Replace non-ISO '-{text[-5:]}' with '{text[-6:]}' in <{localname}>")
-
-    text, n = re.subn(r"(.*\.)(\d{6})\d+(\+.*)", r"\1\2\3", text)
-    if n > 0:
-        log.debug(f"Truncate sub-microsecond time in <{localname}>")
-
-    reader.push(elem, datetime.fromisoformat(text))
+    reader.push(elem, isoparse(text))
 
 
 @end(

--- a/sdmx/tests/reader/test_reader_xml.py
+++ b/sdmx/tests/reader/test_reader_xml.py
@@ -1,4 +1,5 @@
 import re
+from datetime import datetime, timedelta, timezone
 from io import BytesIO
 from itertools import chain
 
@@ -10,6 +11,7 @@ from sdmx.format.xml import qname
 from sdmx.model import Facet, FacetType, FacetValueType
 from sdmx.reader.sdmxml import Reader, XMLParseError
 from sdmx.tests.data import specimen, test_files
+from sdmx.writer.xml import Element as E
 
 
 # Read example data files
@@ -76,8 +78,6 @@ def test_read_ss_xml():
     assert len(TIME_FORMAT.related_to.dimensions) == 5
 
 
-E = etree.Element
-
 # Each entry is a tuple with 2 elements:
 # 1. an instance of lxml.etree.Element to be parsed.
 # 2. Either:
@@ -86,7 +86,25 @@ E = etree.Element
 #   - A string, in which case parsing the element is expected to fail, raising
 #     an exception matching the string.
 ELEMENTS = [
-    # Reader.parse_facet
+    # sdmxml._datetime()
+    (  # with 5 decimal places
+        E(qname("mes:Extracted"), "2020-08-18T00:14:31.59849+05:00"),
+        datetime(2020, 8, 18, 0, 14, 31, 598490, tzinfo=timezone(timedelta(hours=5))),
+    ),
+    (  # with 7 decimal places
+        E(qname("mes:Extracted"), "2020-08-18T01:02:03.4567891+00:00"),
+        datetime(2020, 8, 18, 1, 2, 3, 456789, tzinfo=timezone.utc),
+    ),
+    (  # with "Z"
+        E(qname("mes:Extracted"), "2020-08-18T00:14:31.59849Z"),
+        datetime(2020, 8, 18, 0, 14, 31, 598490, tzinfo=timezone.utc),
+    ),
+    pytest.param(  # with 7 decimal places AND "Z"
+        E(qname("mes:Extracted"), "2020-08-18T01:02:03.4567891Z"),
+        datetime(2020, 8, 18, 1, 2, 3, 456789, tzinfo=timezone.utc),
+        marks=pytest.mark.xfail(raises=XMLParseError),
+    ),
+    # sdmxml._facet()
     (
         E(qname("str:TextFormat"), isSequence="False", startValue="3.4", endValue="1"),
         None,

--- a/setup.cfg
+++ b/setup.cfg
@@ -33,6 +33,7 @@ install_requires =
     lxml >= 3.6
     pandas >= 1.0
     pydantic >= 1.1
+    python-dateutil
     requests >= 2.7
     setuptools >= 41
 setup_requires =

